frontend/exporter/src/id_table.rs

```diff
@@ -0,0 +1,375 @@
+/// This module provides a notion of table, identifiers and nodes. A
+/// `Node<T>` is a `Arc<T>` bundled with a unique identifier such that
+/// there exists an entry in a table for that identifier.
+///
+/// The type `WithTable<T>` bundles a table with a value of type
+/// `T`. That value of type `T` may hold an arbitrary number of
+/// `Node<_>`s. In the context of a `WithTable<T>`, the type `Node<_>`
+/// serializes and deserializes using a table as a state. In this
+/// case, serializing a `Node<U>` produces only an identifier, without
+/// any data of type `U`. Deserializing a `Node<U>` under a
+/// `WithTable<T>` will recover `U` data from the table held by
+/// `WithTable`.
+///
+/// Serde is not designed for stateful (de)serialization. There is no
+/// way of deriving `serde::de::DeserializeSeed` systematically. This
+/// module thus makes use of global state to achieve serialization and
+/// deserialization. This modules provides an API that hides this
+/// global state.
+use crate::prelude::*;
+use std::{
+    hash::{Hash, Hasher},
+    sync::{atomic::Ordering, Arc, LazyLock, Mutex},
+};
+
+/// Unique IDs in a ID table.
+#[derive_group(Serializers)]
+#[derive(Default, Clone, Debug, JsonSchema, Hash, PartialEq, Eq, PartialOrd, Ord)]
+#[serde(transparent)]
+pub struct Id {
+    id: u32,
+}
+
+/// A session providing fresh IDs for ID table.
+#[derive(Default, Debug)]
+pub struct Session {
+    next_id: Id,
+    table: Table,
+}
+
+impl Session {
+    pub fn table(&self) -> &Table {
+        &self.table
+    }
+}
+
+/// The different types of values one can store in an ID table.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum Value {
```

Comment:
Regarding the enum, I was planning to copy my [`type_map`](https://github.com/AeneasVerif/charon/blob/b38710c4655e9eaf91beac02c15030e9cbf87999/charon/src/common.rs#L58) module to hax to make it possible to cache `FullDef<Body>`. We could use it here, it would nicely separate the "caching" part from the "type-generic" part.

Reply:
Yeah, I was thinking about that also, but that required more serde custom serializer and deserializer, so I decided an enum would do the trick for now.
But definitively, that could be a nice improvement :)

Reply:
Actually your thing is neat, it just took me a moment to follow what bit did what. I would find it clearer if there was a separate module for the type-generic map and if the table was a wrapper around it

Reply:
Cool! Just pushed b17c8db8b35e680